extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -0,0 +1,703 @@
+      id: template && template.id,
+    };
+  }
+
+  /**
+   * Create an array for date objects to be consumed by AMP actions and events
+   * or amp-bind.
+   * @param {!moment} startDate
+   * @param {?moment} endDate
+   * @return {!Array<!BindDatesDef>}
+   */
+  getBindDates_(startDate, endDate) {
+    const dates = [];
+    const normalizedEndDate = endDate || startDate;
+    if (!normalizedEndDate.isAfter(startDate)) {
+      return dates;
+    }
+
+    const index = startDate.clone();
+    while (!index.isAfter(normalizedEndDate)) {
+      dates.push(this.getBindDate_(index));
+      index.add(1, 'days');
+    }
+
+    return dates;
+  }
+
+  /**
+   * Formats a date in the page's locale and the element's configured format.
+   * @param {?moment} date
+   * @return {string}
+   */
+  getFormattedDate_(date) {
+    const isUnixTimestamp = this.format_.match(/[Xx]/);
+    return date && (isUnixTimestamp ?
+        date.clone().locale(DEFAULT_LOCALE) : date).format(this.format_);
+  }
+
+  /**
+   * Render a day in the calendar view.
+   * @param {!moment} day
+   */
+  renderDay(day) {
+    const key = day.format(DEFAULT_FORMAT);
+    const mapDay = this.renderedDays_[key];
+    if (mapDay) {
+      return mapDay;
+    }
+
+    const dayTemplateData = this.getDayTemplateData_(day);
+    const templatePromise = this.renderDayTemplate_(day, dayTemplateData);
+    this.renderedDays_[key] = this.renderPromiseIntoReact_(templatePromise);
+    return this.renderedDays_[key];
+  }
+
+  /**
+   * Returns `true` if a day template exists.
+   * @return {boolean}
+   */
+  hasDayTemplate_() {
+    return this.templates_.hasTemplate(this.element, '[date-template]');
+  }
+
+  /**
+   * @param {!Array<!DateTemplateMapDef>} templates
+   * @param {!moment} date
+   * @return {?Element}
+   */
+  getTemplate_(templates, date) {
+    for (let i = 0; i < templates.length; i++) {
+      if (templates[i].dates.contains(date)) {
+        return templates[i].template;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Get the template tag corresponding to a given date.
+   * @param {!moment} date
+   * @return {?Element}
+   */
+  getDayTemplate_(date) {
+    return (
+      this.getTemplate_(this.srcTemplates_, date) ||
+      this.getTemplate_(this.elementTemplates_, date) ||
+      this.srcDefaultTemplate_ ||
+      scopedQuerySelector(this.element, '[date-template][default]')
+    );
+  }
+
+  /**
+   * Render the template that corresponds to the date with its data.
+   * @param {!moment} day
+   * @param {!JsonObject} data
+   */
+  renderDayTemplate_(day, data) {
+    const template = this.getDayTemplate_(day);
+    return this.renderTemplate_(template, data, day.format('D'));
+  }
+
+  /**
+   * Render the info section of the calendar view.
+   */
+  renderInfo() {
+    if (!this.infoTemplatePromise_) {
+      this.infoTemplatePromise_ = this.renderInfoTemplate_();
+    }
+    this.scanForBindings_([this.container_]);
```

Comment:
how about other template types? would it make sense to move this to `renderTemplate_`?